Spectral clustering: support pre-calculated affinity matrix, revamp API

ClaudiaComito opened the pull request:
## Due Diligence
<!--- Please address the following points before setting your PR "ready for review".
--->
- General:
    - [x]  **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [x] unit tests: all split configurations tested
    - [x] unit tests: multiple dtypes tested
    - [ ] benchmarks: created for new functionality
    - [ ] benchmarks: performance improved or maintained
    - [x] documentation updated where needed

## Description

<!--- Include a summary of the change/s.
Please also include relevant motivation and context. List any dependencies that are required for this change.
--->

This PR introduces some changes to the Spectral clustering class (see below), this work in connection with parallelization efforts of the [SCIMES](https://github.com/Astroua/SCIMES) package with @dcolombo


Issue/s resolved: #1740

## Changes proposed:

- change class name to `SpectralClustering` (formerly `Spectral`) to match API of the [current scikit-learn version](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.SpectralClustering.html)
- rename `metric` parameter to `affinity` to match sklearn
- support use of pre-calculated affinity matrix 
- introduced new parameter `eigen_solver`, with options `lanczos` and `zolotarev` (Default: `zolotarev`)  (e.g.  #1697 and #1824 )
- introduced new parameters `n_components`, `random_state` to match scikit-learn API
- output of `__spectral_embedding` is now just the embedding, like for scikit-learn (instead of eigenvalues and eigenvectors);  
- introduced  method `DNDarray.diagonal()` for convenience and to match numpy API
- adapted documentation
- adapted tests
- 

## Type of change
<!--
i.e.
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Documentation update
--->

## Memory requirements
<!--- Compare memory requirements to previous implementation / relevant torch operations if applicable:
- in distributed and non-distributed mode
- with `split=None` and `split not None`

This can be done using https://github.com/pythonprofilers/memory_profiler for CPU memory measurements,
GPU measurements can be done with https://pytorch.org/docs/master/generated/torch.cuda.max_memory_allocated.html.
These tools only profile the memory used by each process, not the entire function.
--->
NA
## Performance
<!--- Compare performance to previous implementation / relevant torch operations if applicable:
- in distributed and non-distributed mode
- with `split=None` and `split not None`

Python has an embedded profiler: https://docs.python.org/3.9/library/profile.html
Again, this will only profile the performance on each process. Printing the results with many processes
may be illegible. It may be easiest to save the output of each to a file.
--->
NA
#### Does this change modify the behaviour of other functions? If so, which?
no

ClaudiaComito committed:
refactor API, add `precomputed` affinity

ClaudiaComito committed:
adapt tests

ClaudiaComito committed:
Merge branch 'main' into features/1740-Support_pre-calculated_affinity_matrix_in_ht_cluster_Spectral

ClaudiaComito committed:
Merge branch 'main' into features/1740-Support_pre-calculated_affinity_matrix_in_ht_cluster_Spectral

ClaudiaComito committed:
introduce eigen_solver parameter

ClaudiaComito committed:
adapt tests

ClaudiaComito committed:
Merge branch 'main' into features/1740-Support_pre-calculated_affinity_matrix_in_ht_cluster_Spectral

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

mrfh92 commented:
see #1824 for the eigh function prototype

mrfh92 commented on heat/cluster/spectral.py:
```diff
@@ -210,7 +223,7 @@ def predict(self, x: DNDarray) -> DNDarray:
             raise NotImplementedError("Not implemented for other splitting-axes")
 
-        _, eigenvectors = self._spectral_embedding(x)
+        _, eigenvectors = self.spectral_embedding(x, self.eigen_solver)
```
Here, the eigenvalue problem is solved another time for the input data of `predict` aka the "test-data". However, I think the same spectral embedding of the "training data" (input of `fit`) should be re-used here.

mrfh92 commented on heat/cluster/spectral.py:
```diff
@@ -168,7 +181,7 @@ def fit(self, x: DNDarray):
             raise NotImplementedError("Not implemented for other splitting-axes")
         # 2. Embed Dataset into lower-dimensional Eigenvector space
-        eigenvalues, eigenvectors = self._spectral_embedding(x)
+        eigenvalues, eigenvectors = self.spectral_embedding(x, self.eigen_solver)
```
Since this is the by far most expensive step of the algorithm, it could make sense to save the result of this as an `_`-attribute of the clusterer-object for later re-usage, e.g., in predict.

mrfh92 reviewed (changes requested):
The API-changes look fine to me 👍 

I have only the two comments directly attached to the code regarding re-usability of the (expensive) results of the eigenvalue decomposition. I would suggest to save them somewhere in the object in order to avoid recomputation.

mrfh92 commented:
addition to my review: just saw that scikit-learn's SpectralClustering does not have a predict on its own but only a fit_predict.

ClaudiaComito committed:
Merge branch 'main' into features/1740-Support_pre-calculated_affinity_matrix_in_ht_cluster_Spectral

ClaudiaComito committed:
add fit_predict

ClaudiaComito committed:
Merge branch 'main' into features/1740-Support_pre-calculated_affinity_matrix_in_ht_cluster_Spectral

ClaudiaComito committed:
Merge branch 'features/1723-Symmetric_eigenvalue_decomposition_and_SVD_based_on_Zolotarev-polar_decomposition' into features/1740-Support_pre-calculated_affinity_matrix_in_ht_cluster_Spectral

ClaudiaComito committed:
support eigh decomposition

ClaudiaComito committed:
add argument n_components

ClaudiaComito committed:
Merge branch 'main' into features/1740-Support_pre-calculated_affinity_matrix_in_ht_cluster_Spectral

ClaudiaComito committed:
Merge branch 'main' into features/1740-Support_pre-calculated_affinity_matrix_in_ht_cluster_Spectral

ClaudiaComito committed:
return n_components eigenval only

ClaudiaComito committed:
Merge branch 'main' into features/1740-Support_pre-calculated_affinity_matrix_in_ht_cluster_Spectral

ClaudiaComito committed:
Merge branch 'main' into features/1740-Support_pre-calculated_affinity_matrix_in_ht_cluster_Spectral

ClaudiaComito committed:
introduce random_state and docs editing

ClaudiaComito committed:
introduce method DNDarray.diagonal

ClaudiaComito committed:
update parameters

ClaudiaComito committed:
set default eigendec to zolotarev

ClaudiaComito committed:
fix tests

ClaudiaComito committed:
expand tests

ClaudiaComito committed:
spectral_emb instead of __spectral_emb

ClaudiaComito committed:
reinstate __spectral_amb

ClaudiaComito committed:
Merge branch 'main' into features/1740-Support_pre-calculated_affinity_matrix_in_ht_cluster_Spectral

ClaudiaComito committed:
Merge branch 'main' into features/1740-Support_pre-calculated_affinity_matrix_in_ht_cluster_Spectral

ClaudiaComito commented on heat/cluster/spectral.py:
```diff
@@ -5,38 +5,43 @@
     Attributes
     ----------
-    n_clusters : int
+    n_clusters : int, default=8
```
```suggestion
    n_clusters : int, optional
```

ClaudiaComito commented on heat/cluster/spectral.py:
```diff
@@ -5,38 +5,43 @@
+    n_components : int, default=None
+        Number of components to use for the embedding. If None, n_clusters is used
+    random_state : int, default=None
+        Random seed for reproducibility. If None, no random seed is set.
```
```suggestion
    random_state : int, optional
        Random seed for reproducibility. If None, no random seed is set.
```

ClaudiaComito commented on heat/cluster/spectral.py:
```diff
@@ -131,25 +151,96 @@ def labels_(self) -> DNDarray:
+        n_components : int, default=8
+            Number of components to use for the embedding
+        eigen_solver : str
+            Eigenvalue decomposition strategy to use. Default is 'randomized' (see documentation in :func:`heat.linalg.reigh`).
```
```suggestion
            Eigenvalue decomposition strategy to use.
```

ClaudiaComito commented on heat/cluster/spectral.py:
```diff
@@ -131,25 +151,96 @@ def labels_(self) -> DNDarray:
+        See Also
+        --------
+        :func:`heat.linalg.lanczos`
+        :func:`heat.linalg.eigh`
```
```suggestion
        :func:`heat.linalg.reigh`
```

ClaudiaComito commented on heat/cluster/spectral.py:
```diff
@@ -131,25 +151,96 @@ def labels_(self) -> DNDarray:
+        --------
+        :func:`heat.linalg.lanczos`
+        :func:`heat.linalg.eigh`
+        :func:`heat.linalg.polar`
```
```suggestion
```

ClaudiaComito commented on heat/cluster/spectral.py:
```diff
@@ -131,25 +151,96 @@ def labels_(self) -> DNDarray:
+            # # We are computing the opposite of the laplacian inplace so as
+            # # to spare a memory allocation of a possibly very large array
+            # L *= -1
+            ####
```
```suggestion
```

ClaudiaComito commented on heat/cluster/spectral.py:
```diff
@@ -159,65 +250,62 @@ def _spectral_embedding(self, x: DNDarray) -> Tuple[DNDarray, DNDarray]:
-                n_oversamples=self.reigh_n_oversamples,
-                power_iter=self.reigh_power_iter,
+            raise NotImplementedError(
+                f"Eigenvalue decomposition method {eigen_solver} is not supported. Supported methods are 'zolotarev' and 'lanczos'."
```
```suggestion
                f"Eigenvalue decomposition method {eigen_solver} is not supported. Supported methods are 'randomized' and 'lanczos'."
```

ClaudiaComito reviewed (commented):
various updates after merging #1457

ClaudiaComito committed:
Update docstring and args defaults

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ClaudiaComito committed:
Update __spectral_embedding docs

ClaudiaComito committed:
Edits

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ClaudiaComito committed:
Remove dead code

ClaudiaComito committed:
Remove dead code

ClaudiaComito committed:
Remove outdated reference to linalg.polar

ClaudiaComito committed:
remove dead code

ClaudiaComito committed:
sklearn API consistency

ClaudiaComito committed:
__spectral_embedding returns embedding only

Change return type from Tuple to DNDarray for the function.

ClaudiaComito committed:
assess n_components before decomposition

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ClaudiaComito committed:
apply standard normalization

ClaudiaComito committed:
Update error message

ClaudiaComito committed:
Refactor comments in spectral.py for clarity

Removed outdated notes about eigenvalues and added comments regarding the Laplacian matrix properties.

brownbaerchen commented on .talismanrc:
```diff
@@ -1,3 +1,10 @@
 threshold: medium
 allowed_patterns:
 - 'uses: [A-Za-z-\/]+@[\w\d]+'
+
+fileignoreconfig:
+- filename: .github/workflows/ci.yaml
+  checksum: 3bf095a6ff388eb7aacb80e3b33ed600b377bbce6a97f01ac1579dc702024a13
+- filename: .github/workflows/ci_full.yaml
+  checksum: 54389cc7b9ddd0010ed99d1194da724dbcc51922358f739a5aa30e65e8e2c0e0
+version: "1.0"
```
```suggestion
```
